### PR TITLE
VQ VAE Algorithm

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
Implementation of [VQ-VAE](https://proceedings.neurips.cc/paper/2017/file/7a98af17e63a0ac09ce2e96d03992fbc-Paper.pdf) algorithm.

### CODE REVIEW

**Comment by hnyu** [comment]
What is the advantage of VQ VAE over the standard VAE+discrete latent ([DiscreteVAE](https://github.com/HorizonRobotics/alf/blob/pytorch/alf/algorithms/vae.py))?

File: alf/algorithms/vq_vae.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+        # [B, n]
+        onehot_encodings = torch.zeros(encoding_indices.shape[0],
+                                       self._num_embeddings)
+
+        onehot_encodings.scatter_(1, encoding_indices, 1)
+
+        # quantization
+        # [B, n] * [n, d] -> [B, d]
+        quantized = torch.matmul(onehot_encodings, self._embedding.weight)
```
Comment (hnyu) [inline alf/algorithms/vq_vae.py]: ``quantized`` can directly be obtained by ``gather`` on ``self._embedding.weight`` using ``encoding_indices``?

File: alf/algorithms/vq_vae.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        """
+        # [B, d]
+        input_embedding, _ = self._encoding_net(inputs)
+
+        # calculate distances
+        # [B, 1] + [n] + [B, n]
+        distances = (
+            torch.sum(input_embedding**2, dim=1, keepdim=True) + torch.sum(
+                self._embedding.weight**2, dim=1) -
+            2 * torch.matmul(input_embedding, self._embedding.weight.t()))
```
Comment (hnyu) [inline alf/algorithms/vq_vae.py]: I feel like this NN isn't as robust as Gumbel-softmax with hard samples.

File: alf/algorithms/vq_vae.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    def __init__(self,
+                 input_tensor_spec,
+                 num_embeddings,
+                 embedding_dim,
+                 encoder_ctor,
+                 decoder_ctor,
+                 optimizer=None,
+                 commitment_loss_weight=1.0,
+                 debug_summaries=False,
+                 name="Vqvae"):
```
Comment (hnyu) [inline alf/algorithms/vq_vae.py]: arguments should have type annotations

File: alf/algorithms/vq_vae.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        super().__init__(debug_summaries=debug_summaries, name=name)
+
+        self._embedding_dim = embedding_dim
+        self._num_embeddings = num_embeddings
+
+        # [n, d]
+        self._embedding = nn.Embedding(self._num_embeddings,
+                                       self._embedding_dim)
+        self._embedding.weight.data.uniform_(-1 / self._num_embeddings,
+                                             1 / self._num_embeddings)
```
Comment (hnyu) [inline alf/algorithms/vq_vae.py]: Can use ``torch.nn.init.uniform_`` on ``self._embedding.weight`` instead

File: alf/algorithms/vq_vae.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                the parameter of encoder_net, decoder_net and embedding vectors.
+            commitment_loss_weight (float): the weight for commitment loss.
+        """
+        super().__init__(debug_summaries=debug_summaries, name=name)
+
+        self._embedding_dim = embedding_dim
+        self._num_embeddings = num_embeddings
+
+        # [n, d]
+        self._embedding = nn.Embedding(self._num_embeddings,
```
Comment (hnyu) [inline alf/algorithms/vq_vae.py]: If you don't need nn.Embedding as a module but just need its weight, then probably it's better to have ``self._embedding = nn.Parameter(...)``.

**Review by hnyu — CHANGES_REQUESTED** [review]

**Comment by Haichao-Zhang** [comment]
> What is the advantage of VQ VAE over the standard VAE+discrete latent ([DiscreteVAE](https://github.com/HorizonRobotics/alf/blob/pytorch/alf/algorithms/vae.py))?

They are indeed closely related, but have some differences in implementation details, e.g. the generation of the discrete code.  Therefore, they might be suitable to be used in different scenarios. For example, VQ-VAE can be used as an online clustering algorithm over the input dataset.

File: alf/algorithms/vq_vae.py (hunk 5 above)
Comment (Haichao-Zhang) [reply alf/algorithms/vq_vae.py]: Change to ``nn.Parameter``

File: alf/algorithms/vq_vae.py (hunk 4 above)
Comment (Haichao-Zhang) [reply alf/algorithms/vq_vae.py]: Switched to use ``torch.nn.init.uniform_``

File: alf/algorithms/vq_vae.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/algorithms/vq_vae.py]: Good point. Simplified this part with indexing operation.

File: alf/algorithms/vq_vae.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/algorithms/vq_vae.py]: Not sure but it could be

File: alf/algorithms/vq_vae.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+        # decoding loss
+        rec = self._decoding_net(quantized_st)[0]
+
+        recon_loss = F.mse_loss(rec, inputs, reduction="none").mean(dim=1)
+
+        if self._debug_summaries and alf.summary.should_record_summaries():
+            with alf.summary.scope(self._name):
+                alf.summary.embedding("vq_embedding",
+                                      self._embedding.weight.detach())
```
Comment (hnyu) [inline alf/algorithms/vq_vae.py]: should remove 'weight'

Comment (Haichao-Zhang) [reply alf/algorithms/vq_vae.py]: Fixed

**Review by hnyu — APPROVED** [review]
